include/boost/multiprecision/cpp_int/addition/add_unsigned.hpp

```diff
@@ -0,0 +1,36 @@
+///////////////////////////////////////////////////////////////
+//  Copyright 2020 John Maddock. Distributed under the Boost
+//  Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at https://www.boost.org/LICENSE_1_0.txt
+
+#ifndef BOOST_MP_ADD_UNSIGNED_HPP
+#define BOOST_MP_ADD_UNSIGNED_HPP
+//
+// Select which actual implementation header to use:
+//
+#ifdef __has_include
+#if __has_include(<immintrin.h>)
+#define BOOST_MP_HAS_IMMINTRIN_H
+#include <immintrin.h>
+#endif
+#endif
+
+#ifdef BOOST_MP_HAS_IMMINTRIN_H
+
+#if defined(__AVX__) && defined(BOOST_HAS_INT128)
```

Comment:
As per intrinsics, I think this should be `__ADX__` ([link](https://clang.llvm.org/doxygen/adxintrin_8h_source.html))

Also please see this before using ADX extensions: [link](https://stackoverflow.com/a/45796549/819979)

Reply:
I've disabled AVX usage for gcc/clang since they don't generate AVX instructions anyway as you point out.  MSVC uses __AVX__ (see https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019), just need to check Intel/linux.

Reply:
Intel IA32 Manual says to check CPUID bits for ADX extensions for ADCX, AODX.

It would be amusing if MSVC defined ability of ADX in Advanced vector extension (AVX, SIMD). SIMD won't work because there's a carry dependency between limbs. AFAIK there's no preprocessor macro for ADX in MSVC.

ADCX would be painful to made to work and would hurt performance in addition.

ADX was originally designed to support ultra fast naive multiplication of 512*512 bits.

Reply:
You're right! :(